app/models/product.rb

```diff
@@ -16,14 +16,27 @@ class Product < ApplicationRecord
     validates :price, numericality: { greater_than_or_equal_to: MINIMUM_PRICE }
     validates :discount_price, numericality: { less_than_or_equal_to: :price, message: "Discount Price can't be greater than Original Price" }, if: :price 
   end
-
+  
+  before_validation :set_discount_price
+  before_destroy :ensure_not_referenced_by_any_line_item
+  before_validation do |product|
+    product.title = 'abc' unless product.title
+  end
+  
   private
 
+  def set_title
+    title = 'abc' unless title || title =~ PRODUCT_TITLE_REGEX
+  end
+
+  def set_discount_price
+    discount_price = price unless discount_price
```

Comment:
what if someone want `discount_price` as `zero`

Reply:
added numericality for ```discount_price``` also